File: interview_prep/algorithm/java/src/main/java/hoa/can/code/ez/Sqrt.java
package hoa.can.code.ez;

//https://leetcode.com/problems/sqrtx/

public class Sqrt {
    public int sqrt(int x) {
        if (x == 0 || x == 1) {
            return x;
        }

        long left = 1;
        long right = x;
        long result = 0;

        while (left <= right) {
            long mid = left + (right - left) / 2;
            long square = mid * mid;

            if (square == x) {
                return (int) mid;
            } else if (square < x) {
                left = mid + 1;
                result = mid;
            } else {
                right = mid - 1;
            }
        }

        return (int) result;
    }
}

File: interview_prep/algorithm/Kotlin/src/main/kotlin/hoa/can/code/Sqrt.kt
package hoa.can.code

class Sqrt {
    fun sqrt(x: Int): Int {
        if (x == 0 || x == 1) {
            return x
        }
        var left: Long = 1
        var right = x.toLong()
        var result: Long = 0
        while (left <= right) {
            val mid = left + (right - left) / 2
            val square = mid * mid
            if (square == x.toLong()) {
                return mid.toInt()
            } else if (square < x) {
                left = mid + 1
                result = mid
            } else {
                right = mid - 1
            }
        }
        return result.toInt()
    }
}